From c7b62d663c886c7448f91fe7e263b63f97a3aa9b Mon Sep 17 00:00:00 2001
From: Garvin Hicking <gh@faktor-e.de>
Date: Wed, 10 Jul 2024 09:56:38 +0200
Subject: [PATCH] [BUGFIX] Fix composer PackageArtifact packagePath resolve for root pkgs

With #103898 the handling of `handleRootPackage()` was streamlined.

This lead to root packages' paths not being set to the base dir
of the composer installation any more.

The missing code is now added again.

Resolves: #104345
Related: #103898
Releases: main, 12.4
Change-Id: I8c66d88ffcacddc2c825964d393d7446db551e68
Reviewed-on: https://review.typo3.org/c/Packages/TYPO3.CMS/+/85198
Tested-by: Nikita Hovratov <nikita.h@live.de>
Tested-by: core-ci <typo3@b13.com>
Reviewed-by: Nikita Hovratov <nikita.h@live.de>
Tested-by: Helmut Hummel <typo3@helhum.io>
Reviewed-by: Helmut Hummel <typo3@helhum.io>
---

diff --git a/Classes/Composer/PackageArtifactBuilder.php b/Classes/Composer/PackageArtifactBuilder.php
index 1a4848a..7aab94d 100644
--- a/Classes/Composer/PackageArtifactBuilder.php
+++ b/Classes/Composer/PackageArtifactBuilder.php
@@ -181,7 +181,7 @@
         $usedExtensionKeys = [];
 
         return array_map(
-            function (array $packageAndPath) use (&$usedExtensionKeys): array {
+            function (array $packageAndPath) use ($rootPackage, &$usedExtensionKeys): array {
                 [$composerPackage, $packagePath] = $packageAndPath;
                 $packageName = $composerPackage->getName();
                 $packagePath = GeneralUtility::fixWindowsFilePath($packagePath);
@@ -210,6 +210,10 @@
                 $usedExtensionKeys[$extensionKey] = $packageName;
                 unset($this->availableComposerPackageKeys[$packageName]);
                 $this->composerNameToPackageKeyMap[$packageName] = $extensionKey;
+                if ($composerPackage === $rootPackage) {
+                    // The root package's path is the Composer base dir
+                    $packagePath = $this->config->get('base-dir');
+                }
                 // Add extension key to the package map for later reference
                 return [$composerPackage, $packagePath, $extensionKey];
             },
